pycaw/utils.py

```diff
@@ -18,22 +22,99 @@
     IID_Empty,
 )
 
+# Define a type variable that extends IUnknown
+COMInterface = TypeVar("COMInterface", bound="IUnknown")
+
 
 class AudioDevice:
     """
     https://stackoverflow.com/a/20982715/185510
     """
 
-    def __init__(self, id, state, properties, dev):
-        self.id = id
-        self.state = state
-        self.properties = properties
+    def __init__(self, dev):
         self._dev = dev
-        self._volume = None
+        self._id = None
+        self._state = None
+        self._properties = None
+        self._interfaces: dict[str, COMInterface] = {}
 
     def __str__(self):
         return "AudioDevice: %s" % (self.FriendlyName)
 
+    def ActivateInterface(self, interface: COMInterface) -> COMInterface:
+        interface_id = interface._iid_
+        if interface_id is None:
+            return None
+
+        interface_name = str(interface_id)
+
+        if interface_name not in self._interfaces:
+            activated = self._dev.Activate(interface_id, comtypes.CLSCTX_ALL, None)
+            interface = activated.QueryInterface(interface)
+            self._interfaces[interface_name] = interface
+
+        return self._interfaces[interface_name]
+
+    @property
+    def id(self):
+        if self._id is None:
+            self._id = self._dev.GetId()
+        return self._id
+
+    @property
+    def state(self) -> AudioDeviceState:
+        if self._state is None:
+            state = self._dev.GetState()
+            self._state = AudioDeviceState(state)
+        return self._state
+
+    @staticmethod
+    def getProperty(self, key):
+        store = self._dev.OpenPropertyStore(STGM.STGM_READ.value)
+        if store is None:
+            return None
+
+        propCount = store.GetCount()
+        for j in range(propCount):
+            pk = store.GetAt(j)
+            name = str(pk)
+
+            if name != key:
+                continue
+
+            try:
+                value = store.GetValue(pk)
+                v = value.GetValue()
+                value.clear()
+                return v
+            except COMError as exc:
+                warnings.warn(
+                    "COMError attempting to get property %r "
+                    "from device %r: %r" % (j, id, exc)
+                )
+                return {}
+
+    @property
+    def properties(self) -> dict:
+        if self._properties is None:
```

Comment:
Are you familiar with guard clauses?
Basically inverting the condition to exit early and deal with the rest of the code one level of indentation lower:
```
if self._properties is not None:
    return self._properties
store = self._dev.OpenPropertyStore(STGM.STGM_READ.value)
...
```
And then do not else, but rest of the code.
That saves one indentation level and the brain doesn't have to stack level of logic when tracking the code.
This is relatively simple code of course, but it's a good practice to generalize to make things easier to follow